chain/blocksync/client.go

```diff
@@ -0,0 +1,443 @@
+	ctx, span := trace.StartSpan(ctx, "GetChainMessages")
+	defer span.End()
+
+	req := &Request{
+		Head:    head.Cids(),
+		Length:  length,
+		Options: Messages,
+	}
+
+	validRes, err := client.doRequest(ctx, req, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return validRes.messages, nil
+}
+
+// Send a request to a peer. Write request in the stream and read the
+// response back. We do not do any processing of the request/response
+// here.
+func (client *BlockSync) sendRequestToPeer(
+	ctx context.Context,
+	peer peer.ID,
+	req *Request,
+) (_ *Response, err error) {
+	// Trace code.
+	ctx, span := trace.StartSpan(ctx, "sendRequestToPeer")
+	defer span.End()
+	if span.IsRecordingEvents() {
+		span.AddAttributes(
+			trace.StringAttribute("peer", peer.Pretty()),
+		)
+	}
+	defer func() {
+		if err != nil {
+			if span.IsRecordingEvents() {
+				span.SetStatus(trace.Status{
+					Code:    5,
+					Message: err.Error(),
+				})
+			}
+		}
+	}()
+	// -- TRACE --
+
+	supported, err := client.host.Peerstore().SupportsProtocols(peer, BlockSyncProtocolID)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get protocols for peer: %w", err)
+	}
+	if len(supported) == 0 || supported[0] != BlockSyncProtocolID {
+		return nil, xerrors.Errorf("peer %s does not support protocol %s",
+			peer, BlockSyncProtocolID)
+		// FIXME: `ProtoBook` should support a *single* protocol check that returns
+		//  a bool instead of a list.
+	}
+
+	connectionStart := build.Clock.Now()
+
+	// Open stream to peer.
+	stream, err := client.host.NewStream(
+		inet.WithNoDial(ctx, "should already have connection"),
+		peer,
+		BlockSyncProtocolID)
+	if err != nil {
+		client.RemovePeer(peer)
+		return nil, xerrors.Errorf("failed to open stream to peer: %w", err)
+	}
+
+	// Write request.
+	_ = stream.SetWriteDeadline(time.Now().Add(WRITE_REQ_DEADLINE))
+	if err := cborutil.WriteCborRPC(stream, req); err != nil {
+		_ = stream.SetWriteDeadline(time.Time{})
+		client.peerTracker.logFailure(peer, build.Clock.Since(connectionStart))
+		// FIXME: Should we also remove peer here?
+		return nil, err
+	}
+	_ = stream.SetWriteDeadline(time.Time{}) // clear deadline // FIXME: Needs
+	//  its own API (https://github.com/libp2p/go-libp2p-core/issues/162).
+
+	// Read response.
+	var res Response
+	err = cborutil.ReadCborRPC(
+		bufio.NewReader(incrt.New(stream, READ_RES_MIN_SPEED, READ_RES_DEADLINE)),
+		&res)
+	if err != nil {
+		client.peerTracker.logFailure(peer, build.Clock.Since(connectionStart))
+		return nil, err
+	}
+
+	// FIXME: Move all this together at the top using a defer as done elsewhere.
+	//  Maybe we need to declare `res` in the signature.
+	if span.IsRecordingEvents() {
+		span.AddAttributes(
+			trace.Int64Attribute("resp_status", int64(res.Status)),
+			trace.StringAttribute("msg", res.ErrorMessage),
+			trace.Int64Attribute("chain_len", int64(len(res.Chain))),
+		)
+	}
+
+	client.peerTracker.logSuccess(peer, build.Clock.Since(connectionStart))
```

Comment:
We should really log a success only after we validate the response. It might be a bit hard to do.